signature/jws/jws.go

```diff
@@ -25,24 +35,374 @@ func NewEnvelope() signature.Envelope {
+	return &signature.Payload{
+		Content:     []byte(e.internalEnvelope.Payload),
+		ContentType: signature.MediaTypePayloadV1}, nil
 }
 
 func (e *envelope) SignerInfo() (*signature.SignerInfo, error) {
-	return nil, fmt.Errorf("not implemented")
+	signInfo := signature.SignerInfo{}
+	if e.internalEnvelope == nil {
+		return nil, signature.SignatureNotFoundError{}
+	}
+
+	// parse protected headers
+	protected, err := parseProtectedHeaders(e.internalEnvelope.Protected)
+	if err != nil {
+		return nil, err
+	}
+	if err := populateProtectedHeaders(protected, &signInfo); err != nil {
+		return nil, err
+	}
+
+	// parse signature
+	sig, err := base64.RawURLEncoding.DecodeString(e.internalEnvelope.Signature)
+	if err != nil {
+		return nil, err
+	}
+	signInfo.Signature = sig
+
+	// parse headers
+	var certs []*x509.Certificate
+	for _, certBytes := range e.internalEnvelope.Header.CertChain {
+		cert, err := x509.ParseCertificate(certBytes)
+		if err != nil {
+			return nil, err
+		}
+		certs = append(certs, cert)
+	}
+	signInfo.CertificateChain = certs
+	signInfo.UnsignedAttributes.SigningAgent = e.internalEnvelope.Header.SigningAgent
+	signInfo.TimestampSignature = e.internalEnvelope.Header.TimestampSignature
+
+	return &signInfo, nil
+}
+
+// sign the given payload and headers using the given signing method and signature provider
+func sign(payload []byte, headers map[string]interface{}, signer signature.Signer) (string, []*x509.Certificate, error) {
+	jsonPHeaders, err := json.Marshal(headers)
+	if err != nil {
+		return "", nil, fmt.Errorf("failed to encode protected headers: %v", err)
+	}
+	protectedRaw := base64.RawURLEncoding.EncodeToString(jsonPHeaders)
+	payloadRaw := base64.RawURLEncoding.EncodeToString(payload)
+	digest := protectedRaw + "." + payloadRaw
+
+	// check external or internal signer
+
+	sigB, err := signer.Sign([]byte(digest))
+	if err != nil {
+		return "", nil, fmt.Errorf("failed to sign digest. error : %v", err)
+	}
+	finalSig := digest + "." + base64.RawURLEncoding.EncodeToString(sigB)
+
+	certs, err := signer.CertificateChain()
+	if err != nil {
+		return "", nil, err
+	}
+	return finalSig, certs, err
+}
+
+func validateCertificateChain(certChain []*x509.Certificate, expectedAlg signature.Algorithm, f func(string) error) error {
+	if len(certChain) == 0 {
+		return f("certificate-chain not present or is empty")
+	}
+
+	err := validateCertChain(certChain)
+	if err != nil {
+		return f(fmt.Sprintf("certificate-chain is invalid, %s", err))
+	}
+
+	keySpec, err := signature.ExtractKeySpec(certChain[0])
+	if err != nil {
+		return f(err.Error())
+	}
+	resSignAlgo := keySpec.SignatureAlgorithm()
+	if resSignAlgo != expectedAlg {
+		return f("mismatch between signature algorithm derived from signing certificate and signing algorithm specified")
+	}
+
+	return nil
+}
+
+// validateCertChain takes an ordered certificate chain and validates issuance from leaf to root
+func validateCertChain(certChain []*x509.Certificate) error {
```

Comment:
See above. We should already have this implementation in the x509 package.

Reply:
Updated.